### ISSUE TITLE
[QST]

### ISSUE THREAD

**Issue by rawnhenry** [NVIDIA/cutlass#2192]
1. Why do we need to include a first batch of padding in example 77? What should the padding size be?
2. How does example 77 handle variable sequence length with TMA? I was skimming the code and expected to see some tensormap replace calls but don't see any.

**Comment by hwu36** [comment]
@v0i0 , would you please take a look?

**Comment by hwu36** [comment]
Rawn, nice to see you here again.

**Comment by v0i0** [comment]
Hey Rawn,

There are a few ways to implement variable sequence length, and this seemed simplest when I wrote this - I'd likely handle it differently today.
The first basic question for var seq len support is whether we want our implementation to be correct no matter what data is in the tensor (think inf or nan), or whether we have control over that (this is often the case).
The second question is whether we have extra memory, either padding in front or in the back.

You are correct that if we want to make no assumptions about the data, and we can't have any padding, then group-gemm-like tensormap.replace is the only option.

If we can have padding, but make no assumption on the data, we can do what is implemented here: We have one "offset dimension" which we use to shift to the beginning of the sequence (with box size 1), and one "masking dimension", which we use to offset s.t. the TMA-based masking is correct. We then initialize the TMA coords such that their mutual effect is neutral.
So instead of having a TMA descriptor of (seqlen, ...) with box (blocking, ...), we have a TMA descriptor of (max_seqlen, max_seqlen) with box (1, blocking), and then index it as (seqlen - max_seqlen, max_seqlen - seqlen). But notice how the first term is negative. So we offset the base ptr by max_seqlen * sequence_stride, and index by (seqlen, max_seqlen - seqlen).
Since TMA descriptors need the entire memory area they address to be addressable, we thus need padding of max_seqlen * seqlen_stride (which is often equal to batch_stride).

Lastly, if we can assume no nan or inf in the data, you could also implement this with only masking during softmax, without all the offset logic or tensormap.replace. Zeroing out the P tensor where appropriate would be sufficient, I believe. This is probably the way I would recommend today.

**Comment by devashishshankar** [comment]
Hi @v0i0, thanks for writing this kernel!


I am finding this very promising for my team's use-case, and I have integrated it with pytorch using python bindings. Variable sequence length case is critical for us, but the front padding is leading to a noticeable perf hit. I can't figure out an efficient way to add the padding to the pytorch tensor avoiding DtoD memcpy. So, I was wondering if it is possible to avoid the padding at the kernel level?

Also had a few questions on your comment for my understanding:

> Since TMA descriptors need the entire memory area they address to be addressable

[Cuda docs](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#tensor-memory-access:~:text=Negative%20indices%20and%20out%20of%20bounds) state that out-of-bound accesses are automatically zeroed out? So why do we need the padding here?

> we have a TMA descriptor of (max_seqlen, max_seqlen)

My understanding is that TMA descriptor should be used for q,k,v reads - so shouldn't it be (max_seq_len, K) instead?

Would be great if you can elaborate if it is possible to remove the padding at the kernel level, and what would be the quickest way of making that change? I should have bandwidth to work on it, but might need some guidance as I'm new to both CUDA and Cutlass. Wondering if something like adding a tensormap.replace call for the first example work?

**Comment by v0i0** [comment]
Hey @devashishshankar!

I would expect masking P to be much easier, i.e. (2), however it will not fix your problem with the output tensor. There, you'd either have to use tensormap.replace, some sort of padding (e.g. to tilesize), or a simt epilogue.

The TMA assumes that all memory that is encoded in the TMA descriptor is mapped (this may not  be spelled out explicitly since you are not going to easily run into it, but it is true at the least on Blackwell; feel free to try it out). Note that your suggestions means that  (seqlen-max_seqlen) is always negative, and will trigger masking to zero from the tma. That's why we fold the max_seqlen offset into the pointer instead.

**Comment by liuqi123123** [comment]
> we can assume no nan or inf in the data, you could also implement this with only masking during softmax, without all the offset logic or tensormap.replace. Zeroing out the P tensor where appropriate would be sufficient, I believe. This is probably the way I would recommend today.


Can someone explain it in more detail?  it's also abstract to me!

**Comment by cloudhan** [comment]
@liuqi123123

If no nan or inf in Q and K, Q@K (contraction along head_dim) always produce valid P matrix in seqlen boundary, for OOB values they non NaN or Inf.

masking during softmax means adding large negative values to the OOB results pre-softmax. The exponentiation (safe softmax) will make them zero.

P@V for OOB seqlen position, P contains zeros. If V has no NaN or Inf. P@V always produce a valid result.

In practice, assumption on OOB values are very bad. Asking the users for padding their tensor is also an extremely bad design choice!
